core/src/main/java/io/substrait/relation/ProtoRelConverter.java

```diff
@@ -490,6 +495,35 @@ private Set newSet(SetRel rel) {
     return builder.build();
   }
 
+  private Rel newHashJoin(HashJoinRel rel) {
+    Rel left = from(rel.getLeft());
+    Rel right = from(rel.getRight());
+    Type.Struct leftStruct = left.getRecordType();
+    Type.Struct rightStruct = right.getRecordType();
+    Type.Struct unionedStruct = Type.Struct.builder().from(leftStruct).from(rightStruct).build();
+    var converter = new ProtoExpressionConverter(lookup, extensions, unionedStruct, this);
+    var builder =
+        HashJoin.builder()
+            .left(left)
+            .right(right)
+            .leftKeys(
+                rel.getLeftKeysList().stream().map(converter::from).collect(Collectors.toList()))
+            .rightKeys(
+                rel.getRightKeysList().stream().map(converter::from).collect(Collectors.toList()))
```

Comment:
I _think_ there might be a bug here. The offsets of the right key list are relative to the right input. However, here the converter is created using the left _and_ right inputs. To illustrate, given two tables like:
```
t1: (a: i64, b: fp64, c:string)
t2: (x: fp64, y: string, z: i64)
```
let's say we want to join on
```
left keys: (0, 1)
right keys: (2, 0)
```
the converter has the following root type
```
(a: i64, b: fp64, c:string, x: fp64, y: string, z: i64)
```
so the left keys `(0,1)` refer to the correct inputs `(a, b)` but the right keys `(2, 0)` refer to `(c, a)` not `(z, x)`.

It would be good to have a test case like this.

Reply:
@vbarua you are correct. Thanks for catching this. It did lack the non-key test and it uncovered this issue. 

I added a converter separately for left and right, and then kept the union-based one for the post join filter. 

Also there are redundant parts in the newly added join test class with respect to the extension roundrip tests. Shall we use another abstraction to keep the common utils?

Reply:
> Also there are redundant parts in the newly added join test class with respect to the extension roundrip tests. Shall we use another abstraction to keep the common utils?

I think it would be worth doing at some point, but I don't think it needs to be part of this PR. I'd rather keep the changeset small.